### PR TITLE
KOGITO-3528 - Erase of WID 'nodes' types

### PR DESCRIPTION
**Pull request by handreyrc**
**JIRA**: [KOGITO-3528](https://issues.redhat.com/browse/KOGITO-3528)

**VS Code**: [plugin](https://drive.google.com/file/d/1tDMO1pi0AAvqn-DsMqHyyn2RTLSsK4GL/view?usp=sharing)
**Business Central**: [WAR file](https://drive.google.com/file/d/1YIrfLSX-uGauZFG0zRn_r5JKGOTCQ5_c/view?usp=sharing)

Hey @romartin, please could you review this PR?  

Thanks

### CODE REVIEW

**Comment by handreyrc** [comment]
jenkins do fdb

**Review by romartin — COMMENTED** [review]
Hey @handreyrc , code looks great! Nothing to append here, everything ok for me, great job :+1:

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java [hunk 1]
```diff
@@ -95,6 +96,21 @@ public void testBuild() {
...
+        // Check for default custom task values
+        assertTrue(customTask.getName().equals("Custom Task"));
+        assertNull(customTask.getTaskType().getRawType());
+        assertTrue(customTask.getExecutionSet().getTaskName().getValue().equals("Service Task"));
+        assertTrue(customTask.getCategory().equals("CustomTasks"));
+        assertTrue(customTask.getDefaultHandler().isEmpty());
+        assertTrue(customTask.getGeneral().getName().getValue().equals("Custom Task"));
+        assertTrue(customTask.getGeneral().getDocumentation().getValue().isEmpty());
+        assertTrue(customTask.getDescription().equals("Custom Task"));
+        assertTrue(customTask.getDataIOSet().getAssignmentsinfo().getValue().isEmpty());
```
Comment (domhanak) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java:111]: @handreyrc does it make sense to also check case where there is also value that is not empty? Let's say we have existing process with WID that has the values for it's assignments provided (complete process). Is there a test case for this option? Apologies if I am missing it, wanted to double check with you.

Comment (romartin) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java:111]: good catch @domhanak ! I also was looking and I don't see tests for non empty data assignments, which is one of the goals for this task, so I'm also looking forward to see it @handreyrc :)

Comment (handreyrc) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java:111]: Hey @domhanak, @romartin, I analyzed your remarks above and the code.

In my understanding, at this point of the execution we are not considering the values coming from the XML yet. These values will be overridden/complemented later. So this class just provides a draft CustomTask object.

This lead us to three scenarios: 
*A new custom task was added from the palette, so a draft CustomTask is created based on the WID file. This is being tested in the  the testBuild() method, the strings WID_PARAMS and WID_RESULTS corresponds to the Inputs and outputs respectively for the Assignments. 
*The WID file exists and we are loading a diagram, so a draft CustomTask for the type is created based on the WID file and some properties are later overridden/complemented by the data coming from the XML. It is the same test scenario as above.
*The WID file does not exist and we are loading a diagram, so an empty CustomTask is created and the properties are later overridden by the data coming from the XML. This is being tested in the testBuildItemNullWID() method. 

Without a WID file we will never have assignments in this scenario indeed. Only the assignments coming from the XML that will be set later in the loading process. 

I think we are covered here.
WDYT?

**Comment by handreyrc** [comment]
jenkins retest this

**Comment by handreyrc** [comment]
jenkins do fdb

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java (hunk 1 above)
Comment (handreyrc) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/workitem/CustomTaskFactoryTest.java:111]: Hey @domhanak, 
I was discussing with @romartin about how could we improve the test coverage in this PR. We agreed that we can go a bit above in the scope of the code changes and write aditional tests ensuring that after unmarshalling the process without the WID file, the CustomTasks are being preserved.
Thanks for the insight @romartin! 

I've just added more tests!
I'm looking forward to your feedback,

Thanks

**Comment by handreyrc** [comment]
jenkins do fdb

**Comment by domhanak** [comment]
@handreyrc Thanks! I checked use cases in VSCode and they worked well. Doing  sanity check on BC..

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/workitem/NoWIDCustomTaskResolutionTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.Silent.class)
+public class NoWIDCustomTaskResolutionTest {
```
Comment (domhanak) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/workitem/NoWIDCustomTaskResolutionTest.java:50]: awesome! @handreyrc

**Review by domhanak — APPROVED** [review]
I have tested VSCode and was not able to reproduce the behavior.

The work item retains its value as expected now, and no information was lost.
Editor retained values for defined field while also retained values added during the time it has not WID present.
Reading the WID back didn't erase the information either.

Business central worked well too. I played around installing and uninstalling the some non default work item definition.

Thanks for the extra test coverage & fix. :) 

AC:
- When there is no WID definition available, node will have default custom icon :heavy_check_mark: 
- When there is no WID definition available, node will retains its assignments and their values :heavy_check_mark: 
- Editor handles edits and saves for WID properly even without WID present :heavy_check_mark: 
- Adding the WID definition of existing node, does not reset the nodes assignments, name, entry/exit scripts, checkboxes (adhoc) to default value. :heavy_check_mark:

**Comment by domhanak** [comment]
@romartin missing approval from your side

**Review by romartin — APPROVED** [review]
nice stuff, thanks @handreyrc !
